src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_util.py

```diff
@@ -3,8 +3,139 @@
+    def _service_adapter(method_name):
+        return '{}#{}.{}'.format(service_model, service_class, method_name)
+
+    return _service_adapter
+
+
+# pylint: disable=too-few-public-methods
+class ServiceGroup(object):
+    def __init__(self, scope, client_factory, service_adapter=None):
+        self._scope = scope
+        self._factory = client_factory
+        self._service_adapter = service_adapter or (lambda name: name)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def group(self, group_name):
+        return CommandGroup(self._scope, group_name, self._factory, self._service_adapter)
+
+
+class CommandGroup(object):
+    def __init__(self, scope, group_name, client_factory, service_adapter=None):
+        self._scope = scope
+        self._group_name = group_name
+        self._client_factory = client_factory
+        self._service_adapter = service_adapter or (lambda name: name)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def command(self, name, method_name):
+        cli_command(self._scope,
+                    '{} {}'.format(self._group_name, name),
+                    self._service_adapter(method_name),
+                    client_factory=self._client_factory)
+
+    def generic_update_command(self, name, getter_op, setter_op):
+        cli_generic_update_command(
+            self._scope,
+            '{} {}'.format(self._group_name, name),
+            self._service_adapter(getter_op),
+            self._service_adapter(setter_op),
+            factory=self._client_factory)
+
+
+# PARAMETERS UTILITIES
+
+def patch_arg_make_required(argument):
+    argument.type.settings['required'] = True
+
+
+def patch_arg_update_description(description):
+    def _patch_action(argument):
+        argument.type.settings['help'] = description
+
+    return _patch_action
+
+
+class ParametersContext(object):
+    def __init__(self, command):
+        self._commmand = command
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def ignore(self, argument_name):
+        register_cli_argument(self._commmand, argument_name, ignore_type)
+
+    def argument(self, argument_name, arg_type=None, **kwargs):
+        register_cli_argument(self._commmand, argument_name, arg_type=arg_type, **kwargs)
+
+    def expand(self, argument_name, model_type, patches=None):
+        # TODO:
+        # two privates symbols are imported here. they should be made public or this utility class
+        # should be moved into azure.cli.core
+        from azure.cli.core.commands import _cli_extra_argument_registry
+        from azure.cli.core.commands._introspection import \
+            (extract_args_from_signature, _option_descriptions)
+
+        from .validators import get_expension_validator
```

Comment:
Yup, this is a dynamical validator, which is used to combine parameters into one argument to fit the need of the SDK APIs.